src/user-session-helper/sdi-user-session-helper.c

```diff
@@ -0,0 +1,206 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/*
+ * This is a helper used to launch snapped user daemons that must run
+ * only when a graphical session is active.
+ *
+ * It does it by checking the current sessions from GDM, and waiting
+ * until a graphical session is detected, in which case it will wait
+ * until one of a series of DBus well-known names do appear in the
+ * session bus to ensure that the session has really started.
+ *
+ * If it has to wait in any of these steps, it will exit directly, so
+ * snapd will have to relaunch it. Only when all these steps are correct
+ * from the beginning will this daemon use execve to replace its code
+ * with the desired executable, passed on the parameters list.
+ *
+ * This helper must be run as a snap user daemon, with `restart-condition:
+ * always` and `restart-delay: 1` at least. Also, it requires the `unity7` and
+ * the `login-session-observe` plugs to be able to access the required
+ * DBus objects.
+ */
+
+#define _GNU_SOURCE
+
+#include "sdi-user-session-helper.h"
+#include <unistd.h>
+
+// wait up to ten seconds before timing out for DBus name appearing
+#define MAX_WAIT 10
+#define NAME_TO_WATCH "org.freedesktop.portal.Desktop"
+
+static gboolean sdi_session_is_desktop(const gchar *object_path) {
+  g_autoptr(OrgFreedesktopLogin1Session) session = NULL;
+  guint32 user;
+  // these values belongs to the session proxy, so they must not be freed
+  GVariant *user_data = NULL;
+  const gchar *session_type = NULL;
+
+  session = org_freedesktop_login1_session_proxy_new_for_bus_sync(
+      G_BUS_TYPE_SYSTEM, G_DBUS_PROXY_FLAGS_NONE, "org.freedesktop.login1",
+      object_path, NULL, NULL);
+  user_data = org_freedesktop_login1_session_get_user(session);
+  if (user_data == NULL) {
+    g_warning("Failed to read the session user data. Forcing a reload.");
+    /* if we can't read the data, we can't know whether we are in a desktop
+     * session or in a text one, so we will assume that we are in a session
+     * desktop to force a reload.
+     */
+    return TRUE;
+  }
+  g_variant_get(user_data, "(uo)", &user, NULL);
+  if (getuid() != user) {
+    // the new session isn't for our user
+    return FALSE;
+  }
+  session_type = org_freedesktop_login1_session_get_type_(session);
+  if (session_type == NULL) {
+    g_warning("Failed to read the session type");
+    return FALSE;
+  }
+  if (g_str_equal("x11", session_type) ||
+      g_str_equal("wayland", session_type) ||
+      g_str_equal("mir", session_type)) {
+    // this is a graphical session
+    return TRUE;
+  }
+  return FALSE;
+}
+
+static gboolean sdi_check_graphical_sessions(Login1Manager *login_manager) {
+  GVariant *sessions = NULL;
```

Comment:
Leaked?

Reply:
Not sure... I'll test it.

Reply:
Checked with valgrind after adding a g_autoptr. It doesn't fail, so yes, it's a leak. Fixing...